classes/decision/types/Decline.inc.php

```diff
@@ -0,0 +1,175 @@
+ */
+
+namespace PKP\decision\types;
+
+use APP\decision\Decision;
+use APP\submission\Submission;
+use Illuminate\Validation\Validator;
+use PKP\context\Context;
+use PKP\decision\steps\Email;
+use PKP\decision\Type;
+use PKP\decision\types\traits\InExternalReviewRound;
+use PKP\decision\types\traits\NotifyAuthors;
+use PKP\decision\types\traits\NotifyReviewers;
+use PKP\decision\Workflow;
+use PKP\mail\mailables\DecisionDeclineNotifyAuthor;
+use PKP\mail\mailables\DecisionNotifyReviewer;
+use PKP\security\Role;
+use PKP\submission\reviewRound\ReviewRound;
+use PKP\user\User;
+
+class Decline extends Type
+{
+    use InExternalReviewRound;
+    use NotifyAuthors;
+    use NotifyReviewers;
+
+    public function getDecision(): int
+    {
+        return Decision::DECLINE;
+    }
+
+    public function getNewStageId(): ?int
+    {
+        return null;
+    }
+
+    public function getNewStatus(): ?int
+    {
+        return Submission::STATUS_DECLINED;
+    }
+
+    public function getNewReviewRoundStatus(): ?int
+    {
+        return ReviewRound::REVIEW_ROUND_STATUS_DECLINED;
+    }
+
+    public function getLabel(?string $locale = null): string
+    {
+        return __('editor.submission.decision.decline', [], $locale);
+    }
+
+    public function getDescription(?string $locale = null): string
+    {
+        return __('editor.submission.decision.decline.description', [], $locale);
+    }
+
+    public function getLog(): string
+    {
+        return 'editor.submission.decision.decline.log';
+    }
+
+    public function getCompletedLabel(): string
+    {
+        return __('editor.submission.decision.decline.completed');
+    }
+
+    public function getCompletedMessage(Submission $submission): string
+    {
+        return __('editor.submission.decision.decline.completed.description', ['title' => $submission->getLocalizedFullTitle()]);
+    }
+
+    public function validate(array $props, Submission $submission, Context $context, Validator $validator, ?int $reviewRoundId = null)
+    {
+        // If there is no review round id, a validation error will already have been set
+        if (!$reviewRoundId) {
+            return;
+        }
+
+        parent::validate($props, $submission, $context, $validator, $reviewRoundId);
+
+        foreach ($props['actions'] as $index => $action) {
+            $actionErrorKey = 'actions.' . $index;
+            switch ($action['id']) {
+                case $this->ACTION_NOTIFY_AUTHORS:
+                    $this->validateNotifyAuthorsAction($action, $actionErrorKey, $validator, $submission);
+                    break;
+                case $this->ACTION_NOTIFY_REVIEWERS:
+                    $this->validateNotifyReviewersAction($action, $actionErrorKey, $validator, $submission, $reviewRoundId);
+                    break;
+            }
+        }
+    }
+
+    public function callback(Decision $decision, Submission $submission, User $editor, Context $context, array $actions)
```

Comment:
Good point. I'm a little nervous about calling it `execute` or a similar `record()` or `recordDecision()`, because it overlaps with `Repo::decision()->add()`. A dev reading the code might expect to be able to do this:

```php
$accept = new Accept();
$accept->execute($decision);
```

In other words, the method is only meant to be called from `Repo::decision()->add()` after that function performs some common actions. It'd be great if I could make it `protected` or something and still invoke it from the repository, but of course that doesn't make sense.

Maybe we could call it `runAdditionalActions()` or something like that?

Reply:
Sure, that's a little lengthy but I'm OK with it.